### DISCUSSION TITLE
Gadgetron image processing chains and client

### DISCUSSION THREAD

**Discussion by xueh2** [gadgetron/gadgetron#1143] [Ideas]
## Background

Gadgetron is well established for image reconstruction (from kspace to image). With the introduction of deep learning and other related techniques, the automated image processing, which was often treated as a separate step after reconstruction, is showing the trend to be merged into reconstruction. A number of examples had been developed on top of gadgetron to perform reconstruction, image analysis with automation, biomarker extraction and reporting. One of these example built by my can be found [here](https://pubs.rsna.org/doi/full/10.1148/ryai.2020200009). In this example, the cardiac perfusion imaging was reconstructed and myocardium was segmented to generate report directly on the recon chain.

This integrated method to build reconstruction and processing as one whole application can have advantages to allow more thorough optimization of entire workflow and design end-to-end clinical applications.

This proposal is to add more components to gadgetron, aiming to better support inline image processing and later reporting. 

### modify to gadgetron_ismrmrd_client

Propose to save the ismrmrd header to the client output under the /header entry. Then the recon results will carry a copy of ismrmrd header.

### image client

The first tool to add is an image client. The function of this client is to take in a ismrmrd h5 file containing images and send one or multiple image series to gadgetron server. Its usage can be:

```
gadgetron_ismrmrd_image_client:
Allowed options:
  -h [ --help ] Produce help message
  -v [ --verbose ]  Verbose mode
  -p [ --port ] arg (=9002) Port
  -a [ --address ] arg (=localhost) Address (hostname) of Gadgetron host
  -f [ --filename ] arg Input file
  -o [ --outfile ] arg (=out.h5) Output file
  -g [ --in-group ] arg (=/image_1/images, /image_2/images) Input data groups, the order to send images will be the same as the input group names
  -G [ --out-group ] arg (=2022-08-04 20:23:28) Output group name
  -c [ --config ] arg (=default_image.xml) Configuration file (remote)
  -C [ --config-local ] arg Configuration file (local)
  -t [ --timeout ] arg (=10000) Timeout [ms]
  -F [ --outformat ] arg (=h5) Out format, h5 for hdf5 and hdr for analyze image
```
This client will first make the connection to the gadgetron, send the ismrmrd protocol header, then send out every image group.

### image buffering and triggering scheme, independent from generic reconstruction mechanism

The gadgetron server receives the ismrmrd protocl header and a stream of images from the client. It needs to buffer, sort and process these images, which is similar to the design of the generic recon chain. In the generic recon chain, the input kspace data was accumulated and triggered. The triggered data is assembled as [RO, E1, E2, CHA, N, S, SLC] array. N and S are two dimensions user can specify.

In the image processing chain, the similar idea of buffering is inherited. But the dimensions for image buffering is proposed to change to include all ismrmrd dimensions [CHA SLC CON PHS REP SET AVE]. The input to the buffering gadget is either a single ismrmrd image (can be either 2D or 3D image) with header and meta data, or an IsmrmrdImageArray. The output of the buffering gadget is an array of images hoNDObjectArray<ImageType>, with the dimensions being [CHA SLC CON PHS REP SET AVE].

User can specify which one or more dimensions are triggering dimension for the image buffering gadget. For example, if the triggering dimension is SLC and AVE, then the buffered images will be sent out once when either slice or average index changes in the incoming image. The resulting shape of image array will be [CHA 1 CON PHS REP SET 1]. If triggering dimension is not set, all images will be buffered and sent out in the close call.

This design allow the chain has two independent triggering scheme. One is in the kspace reconstruction step and another is in the image processing step. For example, the kspace recon may be triggered by slice. The image processing can be triggered by average.

### Examples applications to show the utilities

The example is to build a deep learning denoising chain for the real-time cine application. The image processing chain expects gfactor map images and regular images as the inputs. The output will be denoised images. The client will first send out the gfactor image series and then the image series. The image processing chain will be triggered by slice and denoised images will be returned to the client.

### Other items

- Propose to consolidate the current client with image client by putting shared functions into separate source files. 
- Image client will not support compression
- Image client will still allow both h5 and analyze image outputs
- Image client should support multiple data types (e.g. short, unsigned short, int, float, complex<float>)
- The testing framework needs extension to support image chain.

**Comment by hansenms** [comment]
Is this different from https://github.com/gadgetron/gadgetron/issues/1074

**Comment by xueh2** [comment]
Yes, I will add this thread to a more complete plan and close the other one. Then after everyone comments on the plan, go ahead to build it up.
